File: 1593 Split a String Into the Max Number of Unique Substrings/1593.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        n = len(s)
        self.ans = 1

        def dfs(i, seen):
            if i == n:
                self.ans = max(self.ans, len(seen))
            elif len(seen) + n - i > self.ans:
                for j in range(i+1, n+1):
                    cur = s[i:j]
                    if cur not in seen:
                        dfs(j, seen + [cur])
        
        dfs(0, [])
        return self.ans

File: 1593 Split a String Into the Max Number of Unique Substrings/1593.cpp
class Solution {
public:
    void find(int i, int& ans, unordered_set<string>& seen, string& s){
        int n = s.size();
        if (i == n){
            ans = max(ans, (int) seen.size());
        } else if (seen.size() + n - i > ans){
            for (int j = i+1; j <= n; j++){
                string cur = s.substr(i, j-i);
                if (!seen.contains(cur)){
                    seen.insert(cur);
                    find(j, ans, seen, s);
                    seen.erase(cur);
                }
            }
        }
    }

    int maxUniqueSplit(string s) {
        unordered_set<string> seen;
        int ans = 1;
        find(0, ans, seen, s);
        return ans;
    }
};
